src/sgame/sg_bot_util.cpp

```diff
@@ -27,12 +27,159 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 #include "sg_bot_util.h"
 #include "Entities.h"
 
+static void ListTeamEquipment( gentity_t *self, unsigned int numUpgrades[], size_t numUpgradesSize, unsigned int numWeapons[], size_t numWeaponsSize );
+static unsigned int ListTeamMembers( unsigned int allies[], size_t alliesSize, team_t team );
+
+static const int MIN_SKILL = 1;
+static const int MAX_SKILL = 9;
+static const int RANGE_SKILL = MAX_SKILL - MIN_SKILL;
+
+// computes a percent modifier out of skill level which is easier to work with
+// between 0 and 1
+static float SkillModifier( int botSkill )
+{
+	return static_cast<float>( botSkill - MIN_SKILL ) / RANGE_SKILL;
+}
+
+static float GetMaximalSpeed( gentity_t const *self );
+static float GetMaximalSpeed( class_t cl );
+
+// helper function to make code more readable
+static float GetMaximalSpeed( gentity_t const *self )
+{
+	return GetMaximalSpeed( static_cast<class_t>( self->client->ps.stats[ STAT_CLASS ] ) );
+}
+
+// returns maximum "ground speed" of a class
+// this is, for now, an ugly switch with hard-coded values taken from
+// in-game measurement.
+// Anything better would involve working with physics. This would be
+// very nice, but requires a dive into how physics works.
+static float GetMaximalSpeed( class_t cl )
+{
+	switch( cl )
+	{
+		case PCL_ALIEN_BUILDER0:
+			return 287;
+		case PCL_ALIEN_BUILDER0_UPG:
+			return 287;
+		case PCL_ALIEN_LEVEL0:
+			return 448;
+		case PCL_ALIEN_LEVEL1:
+			return 399;
+		case PCL_ALIEN_LEVEL2:
+			return 383;
+		case PCL_ALIEN_LEVEL2_UPG:
+			return 383;
+		case PCL_ALIEN_LEVEL3:
+			return 383;
+		case PCL_ALIEN_LEVEL3_UPG:
+			return 383;
+		case PCL_ALIEN_LEVEL4:
+			return 353;
+		case PCL_HUMAN_NAKED:
+			return 383;
+		case PCL_HUMAN_LIGHT:
+			return 383;
+		case PCL_HUMAN_MEDIUM:
+			return 367;
+		case PCL_HUMAN_BSUIT:
+			return 367;
+		case PCL_NONE:
+		case PCL_NUM_CLASSES:
+		default:
+			ASSERT( false && "invalid class number provided" );
+	}
+}
+
+template <typename T>
+struct equipment_t
+{
+	int &authorized;
+	T item;
+	int price( void ) const
+	{
+		return BG_Class( item )->price;
```

Comment:
How could this possibly work? BG_Class only takes a class_t as its argument, not a weapon_t or upgrade_t. Likewise, BG_Upgrade only works on an upgrade_t.